### DISCUSSION TITLE
Lolin D32 Pro Flashing

### DISCUSSION THREAD

**Discussion by eboston** [HASwitchPlate/openHASP#235] [Q&A] [answered]
Does anyone have any suggestion on how to program this board?  The programmer never connects.  I have tried pressing the reset button while the programmer tries to connect, but no luck.  I'm thinking I got a bad unit because almost all ESP32 boards I have used in the past have the led flash program on it.  This one didn't flash the LED.

So before I send it back, just wanted to see if anyone new any tricks for it.

**Comment by fvanroie** [comment]
What tool are you using to flash the board? No pressing should be required as this is typically done automatically via the Serial port.
The button is for resetting the board, i.e. when it hangs or you want to manually restart it.

I hardly ever touch the ESP during the flashing, as the upload tool does that. Granted not all boards have this feature, the Lolin D32 Pro V2 does.

On Windows you can use [Tasmota-PyFlasher](https://github.com/tasmota/tasmota-pyflasher/releases/tag/1.0), a simple GUI tool for flashing ESP32 firmware without any installation.

**Reply by eboston** [reply]
I have tried Tasmota-PyFlasher, NodeMCU PyFlasher, and PlatformIO.  They all have the same problem.  The board shows up as a plain USB Serial Port where my other boards report as USB-SERIAL CH320 or something to that nature.

**Reply by fvanroie** [reply]
Sounds like the [CH340C driver](https://learn.sparkfun.com/tutorials/how-to-install-ch340-drivers/all)  is missing on your system. Try installing it first.

**Comment by RiggsUS** [comment, marked as answer]
I don't know if I'm not lookin in the right place but I was looking at https://github.com/HASwitchPlate/openHASP/releases for the LOLIN D32 Pro but did not see a bin for it. I have compiled it with vscode but I just thought I would be lazy and try a precompiled one.

**Reply by fvanroie** [reply]
You are right, thanks for letting us know. Since it is a recommended configuration, I have added `lolin-d32-pro_ili9341` to the build-script.
It will be built automatically on each commit and with the next release it will be included by default.

**Comment by wolffman122** [comment]
So I am running into an issue just a white screen but I can get into the web interface. I have a lolin-d32-pro and the screen is this one [Screen](https://www.amazon.com/gp/product/B08MQLSCRP/ref=as_li_qf_asin_il_tl?ie=UTF8&tag=thingiverse09-20&creative=9325&linkCode=as2&creativeASIN=B08MQLSCRP&linkId=fe9efcc549172bbd31c6bab6b8999fc6l). I followed the guide here [video](https://www.amazon.com/gp/product/B08MQLSCRP/ref=as_li_qf_asin_il_tl?ie=UTF8&tag=thingiverse09-20&creative=9325&linkCode=as2&creativeASIN=B08MQLSCRP&linkId=fe9efcc549172bbd31c6bab6b8999fc6) for wiring, what do I have to change in the build to make this work?  I flashed lolin-d32-pro_ili9341_full_16MB_v0.6.2.bin and tried making a custom bin the [instructions](https://www.thingiverse.com/thing:4873759) here.  Not sure if did this write either never used platformio before.

**Reply by fvanroie** [reply]
Thanks for the info. The problem is that the `lolin-d32-pro_ili9341_full_16MB_v0.6.2.bin` firmware is built for the ILI9341 panel. Since you are using an ILI9488, you'll need to use a build which incorporates the correct drivers.

Can you try `freetouchdeck_full_4MB_v0.6.2.bin` since it has the same panel. Then use the FreeTouchDeck pins as [described here](https://github.com/HASwitchPlate/openHASP/blob/master/user_setups/esp32/freetouchdeck.ini). That should work. The only downside is that only 4MB of the flash partition will be used, but at least you should have an image on the display.

For enabling all features a custom build might be needed, but I'd stick with the FreeTouchDeck for testing and see how far that takes you. If that binary works, an 8MB version can be found in the nightly builds. If needed, we can add a 16MB version too.

**Reply by wolffman122** [reply]
I will give it a shot but how do I change the pins on the bin file?  Sorry again really new at platformio?

**Reply by fvanroie** [reply]
You can't unless you compile a custom firmware with custom driver/pins.
You'll need to rewire the pins to use the FreeTouchDeck firmware on the Lolin D32 Pro... or create a custom binary.

**Reply by wolffman122** [reply]
So I go images on the screen.  My one issue is touch now.  From the screen I have the T_CS pin connected to pin 21 on the board.  Is there any of the other touch pins on the screen that I need to connect to get touch working?

**Reply by fvanroie** [reply]
Great! You need to connect 4 wires to the Touch pins to have SPI communications:

TFT | Label | Lolin | Function
-- | -- | -- | --
10 | T_CLK | 18 | Touch SPI bus clock signal
11 | T_CS | 21 | Touch screen chip select signal, low level enable
12 | T_DIN | 23 | Touch SPI bus input
13 | T_DO | 19 | Touch SPI bus output

SPI MISO, MOSI and SCLK are shared between the touch controller and the LCD controller.
T_IRQ does not need to be connected as it is not used in openHASP.

**Reply by wolffman122** [reply]
I Hate to bring up a dead horse, I figured out my problem with my older ILI9488 didn't have a touch chip on it.  Now I got a new ILI9341 with a touch chip and wanted to get a full list of PIN's for screen and touch but haven't been able to find it.  I flashed the lolin-d32-pro-ili9341_full bin file.

**Reply by fvanroie** [reply]
The `lolin-d32-pro-ili9341` firmware has these pins configured:

MSP2402 Pin| Function            |Lolin Pin|Config Name|Display Pin |
---|---------------------| :---:   |-----------|------------|
1  | 5V/3.3V power input | 5V/3.3V |           | VCC
2  | Module Ground       | GND     |           | GND
3  | Chip Select         | GPIO14  | TFT_CS    | CS
4  | LCD Reset line      | GPIO33  | TFT_RST   | RESET
5  | Data Command control| GPIO27  | TFT_DC    | DC
6  | Data Input          | GPIO23  | TFT_MOSI  | SDI(MOSI)
7  | SPI Clock           | GPIO18  | TFT_SCLK  | SCK
8  | Backlight           | GPIO32   | TFT_BCKL  | LED
9  | Data Output (not used) | GPIO19 or N/C    | TFT_MISO  | SDO(MISO)
10 | Touch SPI Clock     | GPIO18  | TFT_SCLK  | T_CLK
11 | Touch Select        | GPIO12  | TOUCH_CS  | T_CS
12 | Touch Data Input    | GPIO23  | TFT_MOSI  | T_DIN
13 | Touch Data Output   | GPIO19  | TFT_MISO  | T_DO
14 | Touch Interrupt     | N/C     |           | T_IRQ
